samples/pubsub/workerpool/main.go

```diff
@@ -0,0 +1,66 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"math/rand"
+	"time"
+
+	"gocloud.dev/internal/workerpool"
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/mempubsub"
+
+	"golang.org/x/sync/errgroup"
+)
+
+func main() {
+	ctx := context.Background()
+
+	// top is a topic that notifies when a video has been uploaded.
+	// sub is a subscription to this topic.
+	top := mempubsub.NewTopic()
+	sub := mempubsub.NewSubscription(top, time.Hour)
+	defer top.Shutdown(ctx)
+	defer sub.Shutdown(ctx)
+
+	var eg errgroup.Group
+	eg.Go(func() error { return publish(ctx, top) })
+	eg.Go(func() error { return subscribe(ctx, sub) })
+	if err := eg.Wait(); err != nil {
+		log.Printf("workerpool sample: %v", err)
+	}
+}
+
+// publish simulates users uploading videos.
+func publish(ctx context.Context, top *pubsub.Topic) error {
+	for {
+		videoURL := fmt.Sprintf("acmestorage://videos/%d.mpeg", rand.Int())
+		if err := top.Send(ctx, &pubsub.Message{Body: []byte(videoURL)}); err != nil {
+			return fmt.Errorf("sending video upload message: %v", err)
+		}
+		// Simulate some delay between videos being uploaded on an up-and-coming video site.
+		time.Sleep(time.Duration(rand.Intn(10) * 100 * int(time.Millisecond)))
+	}
+	return nil
+}
+
+// subscribe pulls from the subscription and simulates processing the videos.
+func subscribe(ctx context.Context, sub *pubsub.Subscription) error {
+	workerpool.Run(ctx, 10, func(ctx context.Context) workerpool.Task {
```

Comment:
I think it reduces clarity to use `workerpool.Task` here, instead of just `func(context.Context)`.  It's not much shorter, and since you can't use it when you actually declare the function, it just looks like something's mismatched.

I suggest either removing that type, or making it unexported if you want to use it inside workerpool.

Reply:
When I do that it gives the following error:
```
./main.go:67:26: cannot use func literal (type func(context.Context) func(context.Context)) as type func(context.Context) workerpool.Task in argument to workerpool.Run
```

Reply:
I prefer it with the `workerpool.Task` type because it seems like a bit of a confusing stutter without that type:
```
	workerpool.Run(ctx, 10, func(ctx context.Context) func(context.Context) {
...
        })
```
Anyway, this will probably become moot within the sample once I respond to your good point below.

Reply:
Yeah, I noticed that. But you could also define a task to take no args, since it is returned from a method that takes  a context, and it is called with the same context.

Reply:
The reason I made the `Task` funcs take a `Context` was to make it clear that they should respect cancellation. What do you think?

Reply:
Fine as you have it.